### PR TITLE
[blocked by #6997]Consolidate Training End Model Checkpoint

### PR DESCRIPTION
**Pull request by shuyingsunshine21**
## What does this PR do?

**Note: as #6997 will fix the global_step and current epoch for training end, it will be useful for this PR. will rebase after that is checked in.**
Master Issue: #6672 

This is to consolidate the part for model checkpointing at the end of training. 

Currently, we checkpoint based on hook `on_validation_end`, it is 
 1. confusing
 2. might cause bug when checkpoint is based on validation metric but we limit_train_batches which prevents validation loop being called. (see related issue also: #6332) or the scenario when training failed and validation loop not called, validation metric is set as monitor (see related issue #5766)

(Note: for end of each training epoch consolidation, need some dependency cleanup, will be in separate PR)

## What this PR do

* move end of training checkpoint logic from training loop to `model_checkpoint` hook `on_train_end`
* instead of relying on `every_n_val_epochs`, we provide option `trigger_on_train_end` to determine whether checkpoint. By default, it is turned off.
* When `trigger_on_train_end` is turned on, to address the issue when `monitor` value is missing, we relax the condition for checking existence of monitor key at end of training. In such case, we fall back to save last.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by shuyingsunshine21** [comment]
@ananthsub , do you think the following could be better? 

`trigger_on_train_end` should be mutually exclusive with rest of the trigger modes (`every_n_train_steps`, `every_n_val_epochs`, ...).

and for this trigger, we only allow `save_last`.

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -239,6 +243,21 @@ def on_validation_end(self, trainer, pl_module) -> None:
...
+        if not self._trigger_on_train_end:
+            return
+        # as we advance one step at end of training, we use global_step - 1
+        # to avoid saving duplicates
+        trainer.global_step -= 1
+        if (not self._should_skip_saving_checkpoint(trainer) and trainer.checkpoint_connector.has_trained):
+            if self.save_last and self.verbose:
+                rank_zero_info("Saving last checkpoint...")
+            self.save_checkpoint(trainer, is_on_train_end=True)
+        trainer.global_step += 1
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/model_checkpoint.py:259]: @shuyingsunshine21 can this not directly call `self._save_last_checkpoint` ? I think the most common case will be for saving a `last.ckpt` file at the end of training. this way we don't thread through the `is_on_train_end` flag everywhere

@carmocca what do you think? would this go along with https://github.com/PyTorchLightning/pytorch-lightning/issues/6470 ?

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 2]
```diff
@@ -239,6 +243,21 @@ def on_validation_end(self, trainer, pl_module) -> None:
...
+        """
+        if not self._trigger_on_train_end:
+            return
+        # as we advance one step at end of training, we use global_step - 1
+        # to avoid saving duplicates
+        trainer.global_step -= 1
+        if (not self._should_skip_saving_checkpoint(trainer) and trainer.checkpoint_connector.has_trained):
+            if self.save_last and self.verbose:
+                rank_zero_info("Saving last checkpoint...")
+            self.save_checkpoint(trainer, is_on_train_end=True)
```
Comment (shuyingsunshine21) [inline pytorch_lightning/callbacks/model_checkpoint.py:258]: @ananthsub ,

> can this not directly call self._save_last_checkpoint ? I think the most common case will be for saving a last.ckpt file at the end of training. this way we don't thread through the is_on_train_end flag everywhere

we could directly call `self._save_last_checkpoint` by ignoring the topK setup.

One thing to discuss is if `trigger_on_train_end` is set, should we guarantee to save `last.ckpt` even if `save_last` is not set?

Comment (ananthsub) [reply pytorch_lightning/callbacks/model_checkpoint.py:258]: I think we should respect what's set on the callback. The other reason is if we have multiple checkpoint callbacks, we don't need them all to save on train end. We'll configure only one of them to have `save_last=True`

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -239,6 +243,22 @@ def on_validation_end(self, trainer, pl_module) -> None:
...
+        # as we advance one step at end of training, we use global_step - 1
+        # to avoid saving duplicates
+        trainer.global_step -= 1
+        if (not self._should_skip_saving_checkpoint(trainer) and trainer.checkpoint_connector.has_trained):
+            if self.save_last and self.verbose:
+                rank_zero_info("Saving last checkpoint...")
+            monitor_candidates = self._monitor_candidates(trainer)
+            self._save_last_checkpoint(trainer, monitor_candidates)
+        trainer.global_step += 1
+
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/model_checkpoint.py:261]:
```suggestion
    def on_train_end(self, trainer, pl_module) -> None:
        """Save a checkpoint at the very end of training.

        This will only save a checkpoint if `save_last` is also enabled
        as the monitor metrics produced by training or validation steps or end of epochs
        is not guaranteed to be available at this stage.
        """
        if self._should_skip_saving_checkpoint(trainer) or not trainer.checkpoint_connector.has_trained:
            return

        initial_save_last = self.save_last
        if self._save_on_train_end and not self.save_last:
            rank_zero_warn(
                "Requested to save a checkpoint at the end of training but save_last is not set. Temporarily setting save_last=True to save."
            )
            self.save_last = True
        if self.verbose:
            rank_zero_info("Saving last checkpoint...")

        # as we advance one step at end of training, we use global_step - 1
        # to avoid saving duplicates
        trainer.global_step -= 1
        monitor_candidates = self._monitor_candidates(trainer)
        self._save_last_checkpoint(trainer, monitor_candidates)
        trainer.global_step += 1
        self.save_last = initial_save_last

```

what do you think of this? 

also what should happen if save_last is not set to True? should save on train end take precedence and temporarily override it? should we move the save_last check out of _save_last_checkpoint so the property needs to be checked first before we call save_last_checkpoint?

Comment (shuyingsunshine21) [reply pytorch_lightning/callbacks/model_checkpoint.py:261]: i think the original thought is `save_on_train_end` is dependent on `save_last`, so only enabled when `save_last` is set also. What you proposed is to always enable is regardless of `save_last`. To make `save_on_train_end` as an independent triggering, makes sense also.

Comment (ananthsub) [reply pytorch_lightning/callbacks/model_checkpoint.py:261]: @carmocca @awaelchli what do you think?

Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:261]: I think I prefer the current implementation, maybe throwing a warning so people know they should set both.

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 4]
```diff
@@ -186,6 +189,7 @@ def __init__(
         every_n_train_steps: Optional[int] = None,
         every_n_val_epochs: Optional[int] = None,
         period: Optional[int] = None,
+        trigger_on_train_end: bool = False,
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:192]: Shouldn't it default to True so the current behaviour is kept?

**Comment by carmocca** [comment]
Do we need the `trigger_on_train_end` flag?

**Comment by shuyingsunshine21** [comment]
Hi @carmocca , if my understanding is correct, your PR: https://github.com/PyTorchLightning/pytorch-lightning/pull/7724 would include this change. Maybe I could abandon this one?

**Comment by carmocca** [comment]
I opened #7724 to have the full picture of what would be necessary to remove `check_checkpoint_callback`. But once that is validated, it can be split into smaller changes including the ones in this PR (after tweaking).

So let's keep this open for now. You can hold off on updating it and I can hijack it later and do it myself if necessary.

Thanks!

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/callbacks/model_checkpoint.py:192]: @carmocca yes, it could be. we'd also need to make `save_last=True` by default too then right? otherwise we'd run into the warning commented about below
